### PR TITLE
Verify gen-all-sdk-grpc

### LINKED ISSUE #3303: Verify `make gen-all-sdk-grpc` has been run
#3302 exposes an issue in our CI: We don't verify that `make gen-all-sdk-grpc` has been run. We should enforce that any SDK API (proto, etc) changes regenerate the gRPC, and we should enforce we can still build with the new generation.

cc @markmandel

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

/kind cleanup

> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #3303

**Special notes for your reviewer**:

### CODE REVIEW

File: build/includes/sdk.mk [hunk 1]
```diff
@@ -213,3 +213,17 @@ sdk-shell-csharp:
 sdk-publish-csharp: RELEASE_VERSION ?= $(base_version)
 sdk-publish-csharp:
 	$(MAKE) run-sdk-command-csharp COMMAND=publish VERSION=$(RELEASE_VERSION) DOCKER_RUN_ARGS="$(DOCKER_RUN_ARGS) -it"
+
+# difference in sdks before and after gen-all-sdk-grpc target
+test-gen-all-sdk-grpc:
+	mkdir -p includes/tmp
+	cp -r ../sdks includes/tmp
```
Comment (zmerlynn) [inline build/includes/sdk.mk:220]: It's also the `test/sdk` directory.

**Review by zmerlynn — COMMENTED** [review]
Looks good - small piece missing

File: build/includes/sdk.mk (hunk 1 above)
Comment (Kalaiselvi84) [reply build/includes/sdk.mk:220]: I have included the 'test/sdk' directory to compare the differences.

**Comment by Kalaiselvi84** [comment]
`make lint` command is running without error in my local, but here it's failing🤔

**Review by zmerlynn — APPROVED** [review]

**Comment by markmandel** [comment]
Just disabling automerge until we get this to pass, so it doesn't block the auto merge queue.

**Comment by Kalaiselvi84** [comment]
May I modify the `test-gen-all-sdk-grpc` target in the sdk.mk file as shown below?

`test-gen-all-sdk-grpc:
	make gen-all-sdk-grpc
	diff_output=$$(git diff --name-status HEAD -- ../sdks); \
	diff_output_test_sdk=$$(git diff --name-status HEAD -- ../test/sdk); \
	if [ -z "$$diff_output" ] && [ -z "$$diff_output_test_sdk" ]; then \
		echo "No differences found."; \
	else \
		echo "Differences found."; \
		echo "$$diff_output"; \
		echo "$$diff_output_test_sdk"; \
	fi`

**Comment by Kalaiselvi84** [comment]
Is the modification we made to the CI build not performing as anticipated? How might we approach rectifying the build issue?

**Comment by markmandel** [comment]
I'm not 100% sure why this is failing. Going to run it again, see if there's something consistent we can see.

**Comment by markmandel** [comment]
Wrote #3331 to see if I could tell where this is failing, because I'm not seeing it.

**Comment by markmandel** [comment]
@zmerlynn you've been staring at this too, do you see why this is failing? I can't see to see where the conformance tests fall over (I'm also wondering if the retry actually... retries?)

**Comment by zmerlynn** [comment]
> you've been staring at this too, do you see why this is failing? I can't see to see where the conformance tests fall over (I'm also wondering if the retry actually... retries?)

Uhhhhhh.. If I had to take a guess it was rust allegedly failing since we see this stanza several times:
```
Step #27 - "sdk-conformance": for try in `seq 1 5`; do \
Step #27 - "sdk-conformance":   /usr/bin/make run-sdk-conformance-no-build && echo "+++ Success: rust" && break || \
Step #27 - "sdk-conformance":     status=$? && echo "*** Failure: rust, try $try/5"; \
Step #27 - "sdk-conformance": done; (exit $status)
```

But like we actually see multiple lines of:
```
Step #27 - "sdk-conformance": +++ Success: rust
```

which indicates we're ... retrying success?

**Comment by markmandel** [comment]
I hadn't noticed that, but that's a good point. Which seems to be a separate issue, but also, why is it failing now?

File: build/includes/sdk.mk [hunk 2]
```diff
@@ -213,3 +213,20 @@ sdk-shell-csharp:
...
+	diff_output=$$(diff -bBr includes/tmp/sdks ../sdks); \
+	diff_output_test_sdk=$$(diff -bBr includes/tmp/sdk ../test/sdk); \
+	if [ -z "$$diff_output" ] && [ -z "$$diff_output_test_sdk" ]; then \
+		echo "No differences found."; \
+	else \
+		echo "Differences found."; \
+		echo "$$diff_output"; \
+		echo "$$diff_output_test_sdk"; \
+	fi
+	rm -r includes/tmp;
```
Comment (markmandel) [inline build/includes/sdk.mk:232]: A thought: Can we add a line here:

```
git diff
```

Let's see if git sees any differences that aren't picked up by our `diff` execution?

**Comment by zmerlynn** [comment]
I have to image there's something getting perturbed by this PR in particular, because I just glanced at `main` and we're not seeing the same flake. Two things:

1) The actual test itself doesn't seem to be failing the step appropriately. I see:

```
if [ -z "$diff_output" ] && [ -z "$diff_output_test_sdk" ]; then \
	echo "No differences found."; \
else \
	echo "Differences found."; \
	echo "$diff_output"; \
	echo "$diff_output_test_sdk"; \
fi
Differences found.
Only in includes/tmp/sdks/cpp/.build/.install/agones/include/agones: sdk.grpc.pb.h
Only in includes/tmp/sdks/cpp/.build/.install/agones/include/agones: sdk.pb.h
Only in includes/tmp/sdks/cpp/.build/.install/agones/include/google/api: annotations.pb.h
Only in includes/tmp/sdks/cpp/.build/.install/agones/include/google/api: http.pb.h
Only in includes/tmp/sdks/cpp/.build/.install/agones/include/protoc-gen-openapiv2/options: annotations.pb.h
Only in includes/tmp/sdks/cpp/.build/.install/agones/include/protoc-gen-openapiv2/options: openapiv2.pb.h
Only in includes/tmp/sdks/cpp/.build/CMakeFiles/agones.dir/src/agones: sdk.grpc.pb.cc.o
Only in includes/tmp/sdks/cpp/.build/CMakeFiles/agones.dir/src/agones: sdk.pb.cc.o
Only in includes/tmp/sdks/cpp/.build/CMakeFiles/agones.dir/src/google: annotations.pb.cc.o
Only in includes/tmp/sdks/cpp/.build/CMakeFiles/agones.dir/src/google: http.pb.cc.o
Only in includes/tmp/sdks/cpp/.build/CMakeFiles/agones.dir/src/protoc-gen-openapiv2: annotations.pb.cc.o
Only in includes/tmp/sdks/cpp/.build/CMakeFiles/agones.dir/src/protoc-gen-openapiv2: openapiv2.pb.cc.o
diff -bBr includes/tmp/sdks/cpp/include/agones/sdk.grpc.pb.h ../sdks/cpp/include/agones/sdk.grpc.pb.h
[...]

and ..

you already got the other one, which is that we should be cleaning the directoty.
```

File: build/includes/sdk.mk (hunk 2 above)
Comment (zmerlynn) [reply build/includes/sdk.mk:232]: We need to carry some sort of status forward here and fail the step if there's a diff - that's not happening yet. (See my other comment, but just realized there's no status here.)

Comment (Kalaiselvi84) [reply build/includes/sdk.mk]: Updated the target using the git diff and it terminates with a status code of 1, signaling a failure. Can you please confirm if this is correct?

**Comment by Kalaiselvi84** [comment]
Looks the updated code is working as intended,

diff_output=$(git diff --name-status HEAD -- ../sdks); \
diff_output_test_sdk=$(git diff --name-status HEAD -- ../test/sdk); \
if [ -z "$diff_output" ] && [ -z "$diff_output_test_sdk" ]; then \
	echo "No differences found."; \
else \
	echo "Differences found."; \
	echo "$diff_output"; \
	echo "$diff_output_test_sdk"; \
	exit 1; \
fi
Differences found.
M	sdks/cpp/include/agones/sdk.grpc.pb.h
M	sdks/cpp/include/agones/sdk.pb.h
M	sdks/cpp/include/google/api/annotations.pb.h
M	sdks/cpp/include/google/api/http.pb.h
M	sdks/cpp/include/protoc-gen-openapiv2/options/annotations.pb.h
M	sdks/cpp/include/protoc-gen-openapiv2/options/openapiv2.pb.h
M	sdks/cpp/src/agones/sdk.grpc.pb.cc

make: *** [includes/sdk.mk:220: test-gen-all-sdk-grpc] Error 1

File: build/includes/sdk.mk [hunk 3]
```diff
@@ -213,3 +213,17 @@ sdk-shell-csharp:
...
+
+# difference in sdks before and after gen-all-sdk-grpc target
+test-gen-all-sdk-grpc:
+	make gen-all-sdk-grpc
+	diff_output=$$(git diff --name-status HEAD -- ../sdks); \
+	diff_output_test_sdk=$$(git diff --name-status HEAD -- ../test/sdk); \
+	if [ -z "$$diff_output" ] && [ -z "$$diff_output_test_sdk" ]; then \
+		echo "No differences found."; \
+	else \
+		echo "Differences found."; \
```
Comment (markmandel) [inline build/includes/sdk.mk:225]:
```suggestion
		echo "FAILURE: Differences found."; \
```

Just so it's clear that this is bad 😃

**Review by markmandel — COMMENTED** [review]
I didn't actually mean to replace it with git dff, but that works as well!

Okay, so sounds like we need to fix cpp from gen'd gRPC content! 😱

**Comment by zmerlynn** [comment]
> I didn't actually mean to replace it with git dff, but that works as well!

One reason I backed off on suggesting it is that the way @Kalaiselvi84 had it before, we actually catch the case where the repo has a file that is no longer generated. That's actually the case right now, you can see it above in https://github.com/googleforgames/agones/pull/3310#issuecomment-1682893740 .

The only way to do that with the `git diff` method is to start by deleting all of the autogenerated code and then `git diff`. Which is potentially viable if we know exactly what to delete.

**Comment by Kalaiselvi84** [comment]
> > I didn't actually mean to replace it with git dff, but that works as well!
> 
> One reason I backed off on suggesting it is that the way @Kalaiselvi84 had it before, we actually catch the case where the repo has a file that is no longer generated. That's actually the case right now, you can see it above in [#3310 (comment)](https://github.com/googleforgames/agones/pull/3310#issuecomment-1682893740) (which was output from the old way).
> 
> The only way to do that with the `git diff` method is to start by deleting all of the autogenerated code and then `git diff`. Which is potentially viable if we know exactly what to delete.

sure, will skip the `git diff` way and stick with 1st approach including an error status when it finds differences.

File: build/includes/sdk.mk [hunk 4]
```diff
@@ -213,3 +213,21 @@ sdk-shell-csharp:
...
+
+# difference in sdks before and after gen-all-sdk-grpc target
+test-gen-all-sdk-grpc:
+	mkdir -p includes/tmp
+	cp -r ../sdks includes/tmp
+	cp -r ../test/sdk includes/tmp
+	make gen-all-sdk-grpc
+	diff_output=$$(diff -bBr includes/tmp/sdks ../sdks); \
+	diff_output_test_sdk=$$(diff -bBr includes/tmp/sdk ../test/sdk); \
+	if [ -z "$$diff_output" ] && [ -z "$$diff_output_test_sdk" ]; then \
```
Comment (markmandel) [inline build/includes/sdk.mk:225]:
```suggestion
	@if [ -z "$$diff_output" ] && [ -z "$$diff_output_test_sdk" ]; then \
```

I think should work nicely to hide the if statement itself from being printed, so we only see the output.

**Review by markmandel — COMMENTED** [review]
So this looks to be working as expected -- but now we need to go fix the cpp sdk and see what's up there.

**Comment by markmandel** [comment]
Looks like an issue in the bash script:

```
diff_output_test_sdk=$(git diff --name-status HEAD -- ../test/sdk); \
@if [ -z "$diff_output" ] && [ -z "$diff_output_test_sdk" ]; then \
	echo "No differences found."; \
else \
	echo "FAILURE: Differences found."; \
	echo "$diff_output"; \
	echo "$diff_output_test_sdk"; \
	exit 1; \
fi
/bin/sh: 3: Syntax error: "then" unexpected
make: *** [includes/sdk.mk:222: test-gen-all-sdk-grpc] Error 2
```

**Comment by Kalaiselvi84** [comment]
> Looks like an issue in the bash script:
> 
> ```
> diff_output_test_sdk=$(git diff --name-status HEAD -- ../test/sdk); \
> @if [ -z "$diff_output" ] && [ -z "$diff_output_test_sdk" ]; then \
> 	echo "No differences found."; \
> else \
> 	echo "FAILURE: Differences found."; \
> 	echo "$diff_output"; \
> 	echo "$diff_output_test_sdk"; \
> 	exit 1; \
> fi
> /bin/sh: 3: Syntax error: "then" unexpected
> make: *** [includes/sdk.mk:222: test-gen-all-sdk-grpc] Error 2
> ```

In order to address the issue, I've removed the `@` in the if statement

**Comment by Kalaiselvi84** [comment]
The build issue for this PR is being addressed in https://github.com/googleforgames/agones/pull/3349, and this PR will be closed.

**Pull request closed without merging**